Java/src/com/couchbase/litecore/fleece/MDict.java

```diff
@@ -13,92 +13,244 @@
  */
 package com.couchbase.litecore.fleece;
 
-public class MDict extends MCollection {
+import com.couchbase.litecore.SharedKeys;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+public class MDict extends MCollection implements Iterable<String> {
+    private FLDict _dict;
+
+    private Map<String, MValue> _map = new HashMap<String, MValue>();
+
+    private final List<String> _newKey = new ArrayList<String>();
+
+    private long _count = 0;
+
+    /* Constructors */
+
     public MDict() {
-        super(init(), false);
-    }
 
-    public void initInSlot(MValue mv, MCollection parent) {
-        initInSlot(mv._handle, parent._handle);
     }
 
-    public void initInSlot(long hMv, long hParent) {
-        initInSlot(_handle, hMv, hParent);
+    public MDict(MValue mv, MCollection parent) {
+        initInSlot(mv, parent);
     }
 
-    public void initAsCopyOf(MDict mDict, boolean isMutable) {
-        initAsCopyOf(mDict._handle, isMutable);
+    public void initInSlot(MValue mv, MCollection parent) {
+        initInSlot(mv, parent, parent != null ? parent.getMutableChildren() : false);
     }
 
-    public void initAsCopyOf(long hMDict, boolean isMutable) {
-        initAsCopyOf(_handle, hMDict, isMutable);
-    }
+    @Override
+    protected void initInSlot(MValue mv, MCollection parent, boolean isMutable) {
+        super.initInSlot(mv, parent, isMutable);
+        if (_dict != null)
+            throw new IllegalStateException("_dict is not null");
 
-    public long count() {
-        return count(_handle);
+        FLValue value = mv.getValue();
+        if (value != null) {
+            _dict = value.asFLDict();
+            _count = _dict.count();
+        } else {
+            _dict = null;
+            _count = 0;
+        }
     }
 
-    public boolean contains(String key) {
-        return contains(_handle, key);
+    public void initAsCopyOf(MDict d, boolean isMutable) {
+        super.initAsCopyOf(d, isMutable);
+        _dict = d._dict;
+        _map = new HashMap<>(d._map);
```

Comment:
question: Is it enough to do a shallow copy instead of deep copy?

Reply:
This part is not clear. I need your input. Do you think what the following line is doing?

https://github.com/couchbaselabs/fleece/blob/260479eac5b274fae9642135978ab7b7d6edd3c5/Fleece/Mutable/MDict.hh#L55

Note .NET doesn't do copy.

Reply:
http://www.cplusplus.com/reference/unordered_map/unordered_map/operator=/

I believe `_map = d._map;` is case 1. But not sure if it is shallow copy or deep copy.